### PR TITLE
feat(java): Indicate if method param is required

### LINKED ISSUE #365: java: when generating builders indicate for each property if it's required in the docs

### PR DESCRIPTION
**Pull request by dagnir**
This commit enhances javadoc generation for Java classes so that required
method/constructor parameters are indicated in their description in the javadoc.

This makes it easy for customers to see which paramaters must be provided when
interacting with libraries from Java.

Fixes #365

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/UsesInterfaceWithProperties.java [hunk 1]
```diff
@@ -34,6 +36,8 @@ public UsesInterfaceWithProperties(final software.amazon.jsii.tests.calculator.I
 
     /**
      * EXPERIMENTAL
+     * 
+     * @param ext the value to be set. This parameter is required.
```
Comment (bmaizels) [inline packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/UsesInterfaceWithProperties.java]: As discussed for method arguments we don't want to say 'the value to be set'.

**Comment by bmaizels** [comment]
Looks like you removed 'the value to be set' from builder setters. I think we only wanted to remove them from method parameters and constructor args?

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 2]
```diff
@@ -1504,3 +1500,25 @@ function isNullable(optionalValue: spec.OptionalValue | undefined): boolean {
...
+        documentation = summary;
+    } else {
+        documentation = '';
+    }
+
+    if (documentation.length > 0) {
+        if (!documentation.endsWith('.')) {
+            documentation += '.';
+        }
+        documentation += " ";
```
Comment (rix0rrr) [inline packages/jsii-pacmak/lib/targets/java.ts]: This space gets added even if the "required" suffix isn't added later on.

Feels like this method could be simplified to something like:


```ts
function paramJavadoc(name: string, optional?: boolean, summary?: string): string {
  const parts = ['@param', name];
  if (summary) { parts.push(endWithPeriod(summary)); }
  if (!optional) { parts.push('This parameter is required.'); }

  return parts.join(' ');
}
```

File: packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/UnionProperties.java [hunk 3]
```diff
@@ -69,7 +69,7 @@ public Builder bar(software.amazon.jsii.tests.calculator.AllTypes bar) {
 
         /**
          * Sets the value of Foo
-         * @param foo the value to be set
+         * @param foo
```
Comment (rix0rrr) [inline packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/UnionProperties.java]: We are losing something here. Bit of a pity.

Comment (dagnir) [reply packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/UnionProperties.java]: This was unintended as @bmaizels. Will add back in!

**Comment by eladb** [comment]
Please fix PR title and add rationale in description

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 4]
```diff
@@ -1504,3 +1500,18 @@ function isNullable(optionalValue: spec.OptionalValue | undefined): boolean {
...
+    const parts = ['@param', name];
+    if (summary) { parts.push(endWithPeriod(summary)); }
+    if (!optional) { parts.push('This parameter is required.'); }
+
+    return parts.join(' ');
+}
+
+function endWithPeriod(s: string): string {
+    if (!s.endsWith('.')) {
+        return s += '.';
```
Comment (rix0rrr) [inline packages/jsii-pacmak/lib/targets/java.ts]: `return ... += `? Typo?

Comment (dagnir) [reply packages/jsii-pacmak/lib/targets/java.ts]: Thanks!

**Comment by dagnir** [comment]
@eladb Done

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 5]
```diff
@@ -1504,3 +1500,18 @@ function isNullable(optionalValue: spec.OptionalValue | undefined): boolean {
         || (spec.isPrimitiveTypeReference(optionalValue.type)
             && optionalValue.type.primitive === spec.PrimitiveType.Any);
 }
+
+function paramJavadoc(name: string, optional?: boolean, summary?: string): string {
+    const parts = ['@param', name];
+    if (summary) { parts.push(endWithPeriod(summary)); }
+    if (!optional) { parts.push('This parameter is required.'); }
```
Comment (eladb) [inline packages/jsii-pacmak/lib/targets/java.ts]: Can we make this a bit more concise. Something like “(required)”

Comment (dagnir) [reply packages/jsii-pacmak/lib/targets/java.ts]: I actually prefer "This parameter is required.". Yes, it's more verbose, but I think javadocs in libraries tend towards the verbose/descriptive side anyway, or at least do not try to be as short as possible. There is not much precedent for making javadocs more terse/concise that I can tell.

FWIW, others Java SDK team prefer ""This parameter is required." as well.

**Comment by bmaizels** [comment]
Looks good. I've disabled auto-merging so we can get a core member of the CDK team to approve it too.

File: packages/jsii-pacmak/lib/targets/java.ts (hunk 5 above)
Comment (eladb) [reply packages/jsii-pacmak/lib/targets/java.ts]: Ok, cool.

File: packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/DeprecatedClass.java [hunk 6]
```diff
@@ -19,6 +19,8 @@ protected DeprecatedClass(final software.amazon.jsii.JsiiObject.InitializationMo
 
     /**
      * @deprecated this constructor is "just" okay
+     * @param readonlyString This parameter is required.
+     * @param mutableNumber
```
Comment (eladb) [inline packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/DeprecatedClass.java]: shouldn't this say "the value to be set"?

Comment (dagnir) [reply packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/DeprecatedClass.java]: No, per previous comment from @bmaizels, we only want that description for params on builder setters, not constructors/class methods.

**Pull request merged**